lib/srv/app/session.go

```diff
@@ -94,10 +106,25 @@ func (s *Server) newSession(ctx context.Context, identity *tlsca.Identity, app t
 		return nil, trace.Wrap(err)
 	}
 
-	return &session{
-		fwd:          fwd,
-		streamWriter: streamWriter,
-	}, nil
+	// Put the session in the cache so the next request can use it for 5 minutes
+	// or the time until the certificate expires, whichever comes first.
+	ttl := utils.MinTTL(identity.Expires.Sub(s.c.Clock.Now()), 5*time.Second)
+	err = s.cache.set(sess, ttl)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return sess, nil
+}
+
+func (s *Server) closeSession(sess *session) {
+	if err := sess.streamWriter.Close(s.closeContext); err != nil {
+		s.log.Debugf("Failed to close stream writer: %v.", err)
```

Comment:
nit: Use WithError here, like below?

```suggestion
		s.log..WithError(err).Debug("Failed to close stream writer.")
```